docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+```kotlin
+val client = S3Client {
+    region = "us-west-2"
+    disableRequestCompression = true
+    requestMinCompressionSizeBytes = 200
+}
+```
+or
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+}
+```
+```kotlin
+// In config file 
+[default]
+disable_request_compression = true
+request_min_compression_size_bytes = 200
+
+```
+Both of these examples would achieve the same result.
+
+## Testing
+Testing should be done using both unit tests and Smithy HTTP protocol compliance tests. The  SDK unit tests will be used to verify behaviour and ensure it doesn't drift over time (unless intended to). And each AWS HTTP protocol test will be used to test the behavior in a more integrated way in comparison to the unit tests.
+
+AWS JSON 1.0 smithy protocol test for `requestCompression` trait:
+```json
+$version: "2.0"
+
+namespace aws.protocoltests.json10
+
+        use aws.protocols#awsJson1_0
+        use smithy.test#httpRequestTests
+        use smithy.test#httpResponseTests
+
+        apply PutWithContentEncoding @httpRequestTests([
+    {
+        id: "SDKAppliedContentEncoding_awsJson1_0"
+        documentation: "Compression algorithm encoding is appended to the Content-Encoding header."
+        protocol: awsJson1_0
+        params: {
+        "data": """
+                NOTE: REMOVED CONTENTS FOR SAKE OF SPACE. TEST CAN BE FOUND IN REFERENCES SECTION 
+                """
+        }
+        method: "POST"
+        uri: "/"
+        headers: {
+        "Content-Encoding": "gzip"
+        }
+    }
+    {
+        id: "SDKAppendsGzipAndIgnoresHttpProvidedEncoding_awsJson1_0"
+        documentation: """
+        Compression algorithm encoding is appended to the Content-Encoding header, and the
+        user-provided content-encoding is NOT in the Content-Encoding header since HTTP binding
+        traits are ignored in the awsJson1_0 protocol.
+        """
+        protocol: awsJson1_0
+        params: {
+        "encoding": "custom"
+        "data": """
+                NOTE: REMOVED CONTENTS FOR SAKE OF SPACE. TEST CAN BE FOUND IN REFERENCES SECTION 
+                """
+        }
+        method: "POST"
+        uri: "/"
+        headers: {
+        "Content-Encoding": "gzip"
+        }
+    }
+])
+
+@requestCompression(
+    encodings: ["gzip"]
+)
+operation PutWithContentEncoding {
+        input: PutWithContentEncodingInput
+}
+
+@input
+structure PutWithContentEncodingInput {
+    @httpHeader("Content-Encoding")
+    encoding: String
+
+    data: String
+}
+```
+
+Unit tests for interceptor:
+```kotlin
+@Test
+fun testCorrectHeaderIsAdded() = runTest {...}
+
+@Test
+fun testAlgorithmIsAppendedInHeader() = runTest {...}
+
+@Test
+fun testCompressionIsSkipped() = runTest {...}
+
+@Test
+fun testGzipCompression() = runTest {...}
+
+@Test
+fun testHandlesMultipleCompressionAlgorithms() = runTest {...}
+
+@Test
+fun testNoValidCompressionAlgorithm() = runTest {...}
+```
```

Comment:
**Opinion**: Showing these tests adds no value to the design. I think they should be removed.